File: src/main/java/innexgo/hours/SessionService.java
/*
 * Innexgo Website
 * Copyright (C) 2020 Innexgo LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package innexgo.hours;

import java.util.stream.Stream;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

@Transactional
@Repository
public class SessionService {

  @Autowired private JdbcTemplate jdbcTemplate;

  public Session getBySessionId(long sessionId) {
    String sql =
        "SELECT * FROM session WHERE session_id=?";
    RowMapper<Session> rowMapper = new SessionRowMapper();
    Session session = jdbcTemplate.queryForObject(sql, rowMapper, sessionId);
    return session;
  }

  public long nextId() {
    String sql = "SELECT max(ses.session_id) FROM session ses";
    Long maxId = jdbcTemplate.queryForObject(sql, Long.class);
    if(maxId == null) {
      return 0;
    } else {
      return maxId + 1;
    }
  }



  public void add(Session session) {
    session.sessionId = nextId();
    // Add session
    String sql =
        "INSERT INTO session values (?, ?, ?, ?)";
    jdbcTemplate.update(
        sql,
        session.sessionId,
        session.creationTime,
        session.creatorUserId,
        session.courseId);
  }

  public boolean existsBySessionId(long sessionId) {
    String sql = "SELECT count(*) FROM session ses WHERE ses.session_id=?";
    int count = jdbcTemplate.queryForObject(sql, Integer.class, sessionId);
    return count != 0;
  }

  // Restrict sessions by
  public Stream<Session> query(
      Long sessionId,
      Long creationTime,
      Long minCreationTime,
      Long maxCreationTime,
      Long creatorUserId,
      Long courseId,
      long offset,
      long count) {
    String sql =
        "SELECT ses.* FROM session ses"
            + " WHERE 1=1 "
            + (sessionId       == null ? "" : " AND ses.session_id = " + sessionId)
            + (creationTime    == null ? "" : " AND ses.creation_time = " + creationTime)
            + (minCreationTime == null ? "" : " AND ses.creation_time > " + minCreationTime)
            + (maxCreationTime == null ? "" : " AND ses.creation_time < " + maxCreationTime)
            + (creatorUserId   == null ? "" : " AND ses.creator_id = " + creatorUserId)
            + (courseId        == null ? "" : " AND ses.course_id = " + courseId)
            + (" ORDER BY ses.session_id")
            + (" LIMIT " + offset + ", " + count)
            + ";";

    RowMapper<Session> rowMapper = new SessionRowMapper();
    return this.jdbcTemplate.query(sql, rowMapper).stream();
  }
}

File: src/session_service.rs
use super::db_types::*;
use super::utils::current_time_millis;
use tokio_postgres::GenericClient;
use innexgo_hours_api::request;

impl From<tokio_postgres::row::Row> for Session {
  // select * from session order only, otherwise it will fail
  fn from(row: tokio_postgres::row::Row) -> Session {
    Session {
      session_id: row.get("session_id"),
      creation_time: row.get("creation_time"),
      creator_user_id: row.get("creator_user_id"),
      course_id: row.get("course_id"),
    }
  }
}

pub async fn add(
  con: &mut impl GenericClient,
  creator_user_id: i64,
  course_id: i64,
) -> Result<Session, tokio_postgres::Error> {
  let creation_time = current_time_millis();

  let session_id = con
    .query_one(
      "INSERT INTO
       session(
           creation_time,
           creator_user_id,
           course_id
       )
       VALUES($1, $2, $3)
       RETURNING session_id
      ",
      &[&creation_time, &creator_user_id],
    ).await?
    .get(0);

  // return session
  Ok(Session {
    session_id,
    creation_time,
    creator_user_id,
    course_id,
  })
}

pub async fn get_by_session_id(
  con: &mut impl GenericClient,
  session_id: i64,
) -> Result<Option<Session>, tokio_postgres::Error> {
  let result = con
    .query_opt(
      "SELECT * FROM session WHERE session_id=$1",
      &[&session_id],
    ).await?
    .map(|x| x.into());

  Ok(result)
}

pub async fn query(
  con: &mut impl GenericClient,
  props: request::SessionViewProps,
) -> Result<Vec<Session>, tokio_postgres::Error> {
  let results = con
    .query(
      " SELECT ses.* FROM session ses WHERE 1 = 1
        AND ($1::bigint[] IS NULL OR ses.session_id IN $1)
        AND ($2::bigint   IS NULL OR ses.creation_time >= $2)
        AND ($3::bigint   IS NULL OR ses.creation_time <= $3)
        AND ($4::bigint   IS NULL OR ses.creator_user_id = $4)
        AND ($5::bigint   IS NULL OR ses.course_id = $5)
        ORDER BY ses.session_id
        LIMIT $6
        OFFSET $7
      ",
      &[
        &props.session_id,
        &props.min_creation_time,
        &props.max_creation_time,
        &props.creator_user_id,
        &props.course_id,
        &props.count.unwrap_or(100),
        &props.offset.unwrap_or(0),
      ],
    ).await?
    .into_iter()
    .map(|row| row.into())
    .collect();

  Ok(results)
}
